store/sqlstore/thread_store.go

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package sqlstore
+
+import (
+	"database/sql"
+	"github.com/mattermost/gorp"
+	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/mattermost/mattermost-server/v5/store"
+	"github.com/pkg/errors"
+)
+
+type SqlThreadStore struct {
+	SqlStore
+}
+
+func (s *SqlThreadStore) ClearCaches() {
+}
+
+func newSqlThreadStore(sqlStore SqlStore) store.ThreadStore {
+	s := &SqlThreadStore{
+		SqlStore: sqlStore,
+	}
+
+	for _, db := range sqlStore.GetAllConns() {
+		table := db.AddTableWithName(model.Thread{}, "Threads").SetKeys(false, "PostId")
+		table.ColMap("PostId").SetMaxSize(26)
+	}
+
+	return s
+}
+
+func threadSliceColumns() []string {
+	return []string{"PostId", "LastReplyAt", "ReplyCount", "Who"}
+}
+
+func threadToSlice(thread *model.Thread) []interface{} {
+	return []interface{}{
+		thread.PostId,
+		thread.LastReplyAt,
+		thread.ReplyCount,
+		thread.Who,
+	}
+}
+
+func (s *SqlThreadStore) createIndexesIfNotExists() {
+	s.CreateIndexIfNotExists("idx_threads_last_reply_at", "Threads", "LastReplyAt")
+	s.CreateIndexIfNotExists("idx_threads_post_id", "Threads", "PostId")
+}
+
+func (s *SqlThreadStore) Cleanup(postId, rootId, userId string) error {
+	if len(rootId) > 0 {
+		thread, err := s.Get(rootId)
+		if err != nil {
+			if err != sql.ErrNoRows {
+				return errors.Wrap(err, "failed to get a thread")
+			}
+		}
+		if thread != nil {
+			thread.ReplyCount -= 1
+			thread.Who = thread.Who.Remove(userId)
+			if _, err = s.Update(thread); err != nil {
+				return errors.Wrap(err, "failed to update thread")
+			}
+		}
+	}
+	_, err := s.GetMaster().Exec("DELETE FROM Threads WHERE PostId = :Id", map[string]interface{}{"Id": postId})
+	if err != nil {
+		return errors.Wrap(err, "failed to update Threads")
+	}
+	return nil
+}
+
+func (s *SqlThreadStore) UpdateFromPosts(transaction *gorp.Transaction, posts []*model.Post) error {
```

Comment:
Is there any reason to expose the transaction outside the store?

Reply:
Thread store is sort of an addon to post store. Not passing transaction will cause inconsistent behavior

Reply:
@jespino - I didn't like this as well, so I reworked the code a bit to move post-store-related stuff inside Post store, so they are not part of the interface now.